minitswap-router/sources/minitswap-router.move

```diff
@@ -0,0 +1,954 @@
+
+        let pools = minitswap::get_pools(l2_init_metadata);
+        let (_, _, op_bridge_id, ibc_channel, virtual_pool, stableswap_pool) = minitswap::unpack_pools_response(pools);
+
+        let (op_bridge_amount, minitswap_amount, stableswap_amount) = if (option::is_some(&preferred_route)) {
+            let route = option::extract(&mut preferred_route);
+            assert!(route <= 2, error::invalid_argument(EINVALID_ROUTE));
+
+            if (route == OP_BRIDGE) {
+                assert!(is_l1_offered && bridge_out, error::invalid_argument(EINVALID_ROUTE));
+                (offer_amount, 0, 0)
+            } else if (route == MINITSWAP) {
+                assert!(option::is_some(&virtual_pool), error::invalid_argument(EINVALID_ROUTE));
+                (0, offer_amount, 0)
+                // return transfer_fa(account, return_asset, to_sdk(receiver), string::utf8(b"transfer"), ibc_channel, string::utf8(b""))
+            } else {
+                assert!(option::is_some(&stableswap_pool), error::invalid_argument(EINVALID_ROUTE));
+                (0, 0, offer_amount)
+                // return transfer_fa(account, return_asset, to_sdk(receiver), string::utf8(b"transfer"), ibc_channel, string::utf8(b""))
+            }
+        } else {
+            let op_bridge_amount = 0;
+            let minitswap_amount = 0;
+            let stableswap_amount = 0;
+            let remain = offer_amount;
+            let simulation_cache = simple_map::create();
+            let batch_amount = offer_amount / number_of_batches;
+            if (batch_amount == 0) {
+                number_of_batches = 1
+            };
+            let index = 0;
+            while (index < number_of_batches - 1) {
+                (op_bridge_amount, minitswap_amount, stableswap_amount) = find_best_route(
+                    &mut simulation_cache,
+                    op_bridge_amount,
+                    minitswap_amount,
+                    stableswap_amount,
+                    is_l1_offered && bridge_out,
+                    option::is_some(&virtual_pool),
+                    stableswap_pool,
+                    offer_asset_metadata,
+                    return_asset_metadata,
+                    batch_amount,
+                );
+
+                remain = remain - batch_amount;
+                index = index + 1;
+            };
+
+            (op_bridge_amount, minitswap_amount, stableswap_amount) = find_best_route(
+                &mut simulation_cache,
+                op_bridge_amount,
+                minitswap_amount,
+                stableswap_amount,
+                is_l1_offered && bridge_out,
+                option::is_some(&virtual_pool),
+                stableswap_pool,
+                offer_asset_metadata,
+                return_asset_metadata,
+                remain,
+            );
+
+            (op_bridge_amount, minitswap_amount, stableswap_amount)
+        };
```

Comment:
there are any reasons last `find_best_route` separately instead of including it in the loop?